crates/util/src/paths.rs

```diff
@@ -93,6 +97,78 @@ impl<T: AsRef<Path>> PathExt for T {
 
         self.as_ref().file_name()?.to_str()?.split('.').last()
     }
+
+    /// Returns a `SanitizedPathBuf` with a cleaned-up version of the current path,
+    /// or an error if the path is empty on Windows.
+    ///
+    /// On Windows, this method removes the `\\?\` prefix from the path. On non-Windows
+    /// platforms, it directly converts the path reference into a `PathBuf`.
+    ///
+    /// Generally, this method should be used when the path is to be displayed to the user,
+    /// and the path should be returned from `canonicalize`.
+    fn sanitized_pathbuf(&self) -> anyhow::Result<SanitizedPathBuf> {
+        #[cfg(target_os = "windows")]
+        return Ok(SanitizedPathBuf(PathBuf::from(
+            self.sanitized_pathbuf_string()?,
+        )));
+        #[cfg(not(target_os = "windows"))]
+        return Ok(SanitizedPathBuf(self.as_ref().to_path_buf()));
+    }
+
+    /// Returns a sanitized string representation of the current path, or an error
+    /// if the path is empty on Windows.
+    ///
+    /// On Windows, this method removes the `\\?\` prefix. On other platforms,
+    /// it simply converts the path to a string.
+    ///
+    /// Generally, this method should be used when the path is to be displayed to the user,
+    /// and the path should be returned from `canonicalize`.
+    fn sanitized_pathbuf_string(&self) -> anyhow::Result<String> {
+        let path_string = self.as_ref().to_string_lossy();
+        #[cfg(target_os = "windows")]
+        {
+            let trimmed_path_string = path_string.trim_start_matches("\\\\?\\");
+            if trimmed_path_string.is_empty() {
+                anyhow::bail!("File path is empty!");
+            }
+            return Ok(trimmed_path_string.to_string());
+        }
+        #[cfg(not(target_os = "windows"))]
+        return Ok(path_string.to_string());
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Hash)]
+pub struct SanitizedPathBuf(PathBuf);
```

Comment:
I've been thinking of something else, more like a state where we have both a proper path and a proper data for displaying:

```rs
struct SanitizedPathBuf {
    original: PathBuf,
    #[cfg(win)]
    trimmed: PathBuf,
}

impl std::fmt::Display for SanitizedPathBuf {
  // #[cfg(not(win))] and #[cfg(win)]
}

impl std::convert::TryFrom for SanitizedPathBuf {
  type Error = anyhow::Error;

  // #[cfg(not(win))] and #[cfg(win)]
}

// Or should we replace it with a custom `impl SlashCommandEntry`?
// Technically, we still need two paths to return, the canonicalized one and the trimmed one?
// In this API, I presumed it's the trimmed (noop trimmed on Unix) one.
impl AsRef<Path> for SlashCommandEntry {
  // #[cfg(not(win))] and #[cfg(win)]
}
```

then no Deref [semi-antipatterns](https://rust-unofficial.github.io/patterns/anti_patterns/deref.html) are needed.

With this, we can convert things on the spot with that new constructor that might return errors.
No other ways around that constructor — there should be no way to sanitize without the error possibility.

Reply:
I'll give it another try and see what I can come up with, really thx!